File: Graphs/calc_equation.py
from collections import defaultdict
from typing import List


class CalcEquation:
    def calcEquation(
        self, equations: List[List[str]], values: List[float], queries: List[List[str]]
    ) -> List[float]:
        def dfs(node: str, target: str, seen: set[str]) -> float:
            if node == target:
                return 1

            for neighbor in graph[node]:
                if neighbor not in seen:
                    seen.add(neighbor)
                    result = dfs(neighbor, target, seen)
                    if result != -1:
                        weight = graph[node][neighbor]
                        return result * weight

            return -1

        # Pre-process graph
        graph = defaultdict(dict)
        for i in range(len(equations)):
            num, denom = equations[i]
            graph[num][denom] = values[i]
            graph[denom][num] = 1 / values[i]

        answers = []
        for origin, dest in queries:
            answers.append(
                -1 if origin not in graph else dfs(origin, dest, set([origin]))
            )

        return answers

# Time Complexity: O(q * ( n + e )) where q is the number of queries, 
# n is the number of unique variables/nodes, and e is the number of equations/edges.
# In the worst case, we may have to visit all variables and equations for each query.
# Space Complexity: O(n + e) for the graph representation, 
# and O(n) for the seen set in the worst case of a deep recursion.

File: Graphs/calcEquation.ts
const calcEquation = (equations: string[], values: number[], queries: string[][]): number[] => {
  const dfs = (num: string, denom: string, res: number, seen: Set<string>): number => {
    if (num === denom) {
      return res;
    }
    seen.add(num);
    for (const [neighbor, weight] of neighbors.get(num)!) {
      if (!seen.has(neighbor)) {
        const ans = dfs(neighbor, denom, res * weight, seen);
        if (ans !== -1) {
          return ans;
        }
      }
    }
    return -1;
  }

  const neighbors: Map<string, [string, number][]> = new Map();
  for (let i = 0; i < equations.length; i++) {
    const [var1, var2] = equations[i];
    const val = values[i];
    if (!neighbors.get(var1)) neighbors.set(var1, []);
    if (!neighbors.get(var2)) neighbors.set(var2, []);
    neighbors.get(var1)!.push([var2, 1 / val]);
    neighbors.get(var2)!.push([var1, val])
  }

  const solutions = []
  for (const [num, denom] of queries) {
    if (!neighbors.has(num) || !neighbors.has(denom)) {
      solutions.push(-1);
    } else {
      solutions.push(dfs(num, denom, 1, new Set([num])));
    }
  }

  return solutions;
}

// Time complexity: O(q * (V + E)) where q is the number of queries, 
// V is the number of variables, and E is the number of equations. 
// This is because for each query, 
// we may have to traverse the entire graph in the worst case.
// Space complexity: O(V + E) for the graph representation 
// and the recursion stack in the worst case.